### DISCUSSION TITLE
Using "gtk3-classic" makes font and colours in claws-mail's folder view suddenly change during runtime. Using default "gtk3" does not show this issue.

### DISCUSSION THREAD

**Discussion by dreirund** [lah7/gtk3-classic#109] [Help & Support] [answered]
Ahoj,

With `gtk3-classic` I encounter the following strange issue with `claws-mail`. With `gtk3` it does not occur.

The issue is:

When I start `claws-mail`, it renders its font in the folder list "properly". Later on, while running, the rendering changes to another font and other colours.

I can reproducibly trigger the change as follows, for example:

1. Start `claws-mail` (-> everything looks as it should).
2. Right-click on the account in the folder list.
3. Select "Create new Folder" -> At this very moment the font and colours change.

The terminal output from `claws mail --debug` does not show anything when doing the action 3. When doing 2. (i.e. right-click on the account in the folder view) it shows  
```
folderview.c:2286:newly selected 0x5591dfe143c0, opened 0x5591dfe143c0
folderview.c:2290:TIMING folderview_selected : 0s000ms
folderview.c:2286:newly selected 0x5591dfe51c70, opened 0x5591dfe143c0
folderview.c:2312:TIMING folderview_selected : 0s000ms
```

I attach screenshots which document this behaviour; personal information  I have censored (all the yellow stuff is manually censored).  
I also attach screenshots of the font, theme and colour settings.

## Used Software:

* [claws mail](http://claws-mail.org/): Current build is carried out on claws mail git checkout from [git://git.claws-mail.org/claws.git](git://git.claws-mail.org/claws.git) with commit hash `2761674cb1c1c82a1ed52793dbdd5a7cb32466dd`, latest commit (according to `git log -n1`) from 2024-02-29, `git describe --tags` shows `4.2.0-65-g2761674cb`.
* GTK3-classic: 3.24.41 ("gtk3-classic" from https://github.com/lah7/gtk3-classic, build from the AUR package at https://aur.archlinux.org/packages/gtk3-classic).
  -Issue does _not occur_ with my standard operating system's GTK3!
* Operating System: [Artix GNU/Linux](http://artixlinux.org/) (rolling release, without systemd).
* Window Manager: xfwm4 (no desktop session, a plain `.xinitrc`-session).

I build claws mail as follows:

<details><summary>claws mail build procedure (click to show):</summary>

```
git clone git://git.claws-mail.org/claws.git claws-mail # Use master branch which is GTK3
cd claws-mail
NOCONFIGURE=1 ./autogen.sh
./configure \
  --prefix=/usr \
  --disable-static \
  --enable-shared \
  --enable-nls \
  --enable-manual \
  --enable-libsm \
  --enable-ipv6 \
  --enable-gnutls \
  --enable-enchant \
  --enable-crash-dialog \
  --disable-generic-umpc \
  --enable-compface \
  --enable-pthread \
  --enable-startup-notification \
  --enable-dbus \
  --enable-ldap \
  --enable-jpilot \
  --disable-networkmanager \
  --enable-libetpan \
  --disable-valgrind \
  --disable-alternate-addressbook \
  --enable-svg \
  --enable-deprecated \
  --enable-acpi_notifier-plugin \
  --enable-address_keeper-plugin \
  --enable-archive-plugin \
  --enable-att_remover-plugin \
  --enable-attachwarner-plugin \
  --enable-bogofilter-plugin \
  --enable-bsfilter-plugin \
  --enable-clamd-plugin \
  --enable-dillo-plugin \
  --enable-fancy-plugin \
  --enable-fetchinfo-plugin \
  --enable-gdata-plugin \
  --enable-libravatar-plugin \
  --enable-litehtml_viewer-plugin \
  --enable-mailmbox-plugin \
  --enable-managesieve-plugin \
  --enable-newmail-plugin \
  --enable-notification-plugin \
  --enable-pdf_viewer-plugin \
  --enable-perl-plugin \
  --enable-python-plugin \
  --enable-pgpcore-plugin \
  --enable-pgpmime-plugin \
  --enable-pgpinline-plugin \
  --enable-rssyl-plugin \
  --enable-smime-plugin \
  --enable-spamassassin-plugin \
  --enable-spam_report-plugin \
  --enable-tnef_parse-plugin \
  --enable-vcalendar-plugin \
  --enable-demo-plugin
make
cd tools
make
cd ..
```
</details>

## Screenshots:

* At startup -- everything as it should:  
  ![01_-_at_startup](https://github.com/lah7/gtk3-classic/assets/1590519/0106caa1-cbfa-447a-a24c-6b257523f329)
* Font ant colours in the folder list messed up:  
  ![02_-_messed-up_colours_and_font_in_folderlist](https://github.com/lah7/gtk3-classic/assets/1590519/0f792922-f128-4309-a6cd-cc788ea4e4c2)
* Settings:  
  ![03_-_fontsettings](https://github.com/lah7/gtk3-classic/assets/1590519/51733234-9db7-4c4c-9014-06859c308166) ![04_-_themesetting](https://github.com/lah7/gtk3-classic/assets/1590519/a9eb0102-fcd9-46dc-9aaa-433671376293) ![05_-_colour-labels](https://github.com/lah7/gtk3-classic/assets/1590519/ffc41a86-13d4-468a-ab0d-12affb034836) ![06_-_colour_others](https://github.com/lah7/gtk3-classic/assets/1590519/1b08a963-d212-4d14-bc65-feb26d90f68f)

---

Regards!

**Comment by lah7** [comment]
Had a quick look. While not an exact replica of your environment, `claws-mail` doesn't inhibit the weird font as bad under regular Arch/KDE (stock Adwaita theme) here, including under the _built from source_ version.

Funnily enough, I had wrote up a bunch of stuff and then finally noticed the **bold text** in the folder list would get lost when "Create folder" dialog opened up, but doesn't happen under GTK 3... same symptom.

<details><summary>Original comment</summary>
I see these GTK errors when the program starts, not sure if they're related:
<pre>(claws-mail:36544): Gtk-WARNING **: 22:54:56.138: Negative content width -3 (allocation 17, extents 10x10) while allocating gadget (node button, owner GtkButton)

(claws-mail:36544): Gtk-WARNING **: 22:54:56.138: Negative content width -3 (allocation 17, extents 10x10) while allocating gadget (node button, owner GtkButton)</pre>

Judging that the font goes funny after the first interaction, I wonder if it's to do with the GTK theme and/or the installed system fonts. I noticed that here I have only have `Sans Regular` but no `Sans`.

What's the name (and package) that provides that GTK theme? _Looks like Redmond._ If you have a `~/.config/gtk-3.0/`, try moving any CSS modifications out in case they're introducing the bug. For instance, `gtk3-classic` patches bring back the alternating colour rows [but require some additional CSS](https://github.com/lah7/gtk3-classic/wiki/Treeview:-Alternating-Colours-CSS) - although Claw Mail must introduce their own alternating tree view.

Claw Mail's theme also said "Default internal theme" here, so I guess that mean it inherits whatever GTK+3 theme is configured to use. Running `claws-mail --debug` shows the same thing when right clicking - I don't think it's an application issue:
<pre>
folderview.c:2286:newly selected 0x5f961d5173a0, opened (nil)
folderview.c:2312:TIMING folderview_selected : 0s000ms
</pre>
</details>

There is [a script in this repository](https://github.com/lah7/gtk3-classic/blob/master/scripts/test-exclude-each-patch.sh) to semi-automate build a bunch of `gtk3-classic` packages that exclude each patch, to identify which one is the cause of the problem ☕

**Comment by lah7** [comment, marked as answer]
After waking up from [yesterday's comment](https://github.com/lah7/gtk3-classic/discussions/109#discussioncomment-8713716), I suspected [`GTK_BACKDROP`](https://github.com/lah7/gtk3-classic?tab=readme-ov-file#appearance), since that patch alters the controls for inactive windows. Turns out it's not that one, but you may need to set [`GTK_CSD=0`](https://github.com/lah7/gtk3-classic?tab=readme-ov-file#client-side-decorations-only-on-xorg) when you run `claws-mail`.

    export GTK_CSD=0
    claws-mail

or

    GTK_CSD=0 claws-mail

This is how it compares - the colour/font doesn't change when using `GTK_CSD=0` to "restore CSDs from the original app". Despite the name/purpose, the app isn't presenting itself as a CSD. Hopefully this'll fix it in your environment? 🤞🏼 

### Before
https://github.com/lah7/gtk3-classic/assets/13032135/cca9ea7a-8022-43c1-85ce-180cdc854039

### After
https://github.com/lah7/gtk3-classic/assets/13032135/de51dd19-e8b2-406c-ac60-d80cf8a7effc

**Reply by dreirund** [reply]
> *but you may need to set [`GTK_CSD=0`](https://github.com/lah7/gtk3-classic?tab=readme-ov-file#client-side-decorations-only-on-xorg) when you run `claws-mail`*

Yes, setting the environment variable `GTK_CSD=0` did solve the issue for me.

But still, this seems just to be a symptom hunt, not a general bugfix, since `gtk3-classic` is designed to work without client side decorations and having to set `GTK_CSD=0` (even if the software in question does not use client side decorations) goes against the objective of `gtk3-classic`, so we still might have an issue with `gtk3-classic` here?

Regards!

**Reply by lah7** [reply]
I agree it's like a symptom hunt, but sort of expected. Developers who write GTK applications can't anticipate weird behaviour from a modified GTK 3 toolkit.

Without `GTK_CSD` set (and X11 only), the code in [csd__disabled-by-default.patch](https://github.com/lah7/gtk3-classic/blob/8e19e23c9fe46801d0b0ac236f038de69323b00b/csd__disabled-by-default.patch) gets executed. You could be right, there might be something deeper going on that impacts how Claws Mail renders the widget when we're disabling client decorations.

I wouldn't say it's against _the objective_, it is what it is. My aim is to be on stand by to accept patches and keep them rolling whenever a new GTK 3 is released. I think the original patch author made a good choice to have this "feature" togglable through an environment variable, for these kinds of scenarios.

**Comment by dreirund** [comment]
Ahoj,

I still experienced two times that font and colour got messed up despite
`GTK_CSD=0`. Not when selecting create new folder, but at some point
during normal `claws-mail` usage.

I have not yet pinned down when exactly it is triggered.

Regards!

**Comment by daemonspudguy** [comment]
2 years into the future and it gets weirder if you switch from a light to a dark theme while Claws is running or vice versa.
<img width="368" height="675" alt="image" src="https://github.com/user-attachments/assets/5e99b28e-d583-4e65-98e2-318d07788d0c" />
<img width="364" height="624" alt="image" src="https://github.com/user-attachments/assets/85f1da88-fb6d-46a0-b68e-8d8f8164cb2b" />
I have absolutely no clue what could possibly be causing this behavior.

**Reply by dreirund** [reply]
Those "white" and with different font rendered folder names I also have in light theme (see original report). They occur not at startup of claws mail, but after some time. (And sometimes also disappears again.) I don't know if it is `gtk3-classic` related or not.

Regards!
